Sprint 16

maxim-papier opened the pull request:
Cлучайно намёрджил в `main`, вместо `sprint_16` раньше времени. Причём, две ветки #17 #18. Последний пул откатил, а предыдущий не смог. Таким образом, часть 16-го спринта, касающаяся онбординга уже влита в `main`. 

Приношу свои извинения за путаницу.

maxim-papier committed:
Refactor category to mvvm (#20)

* Onboarding (#17)

*
- Added onboarding and attempting to implement it using MVVM
- Added log service
- Added app icon

Known issues:
- PageControl switches incorrectly
- Can't press the button to close onboarding
- The fact of closing onboarding is not saved in UserDefaults

* 
Onboarding will not be shown once it has been completed
- Implemented a service to save the onboarding status in UserDefaults
- Improved the separation of responsibilities between the Onboarding View and ViewModel
- The button to complete onboarding has been implemented
- The layout of the onboarding page elements has been corrected according to the Figma layout

* Fixed blinking issue when Onboarding appears.

* Renamed Stores classes and Models to more intuitive names

* Further optimized `CategoryStore` for MVVM approach:
- Removed all unused methods
- Converted all conversion methods to private
- Renamed public methods for better clarity.

* Refactored `isTrackerReadyToBeCreated` method and got rid off all print()'s

* Added `CategoryCellViewModel`, `CategoryViewModel` and pre-refactored `CategoryCell`

* Refactored the CategoryVC scene according to the MVVM pattern

Known issues:
Occasionally, separators disappear when a category is selected
The interface does not update after a category is added
I'm not satisfied with the behavior of the .ready / .add button

* Implemented a functional `CategoryView` following the MVVM approach.
- Fixed issues with separators
- Table now refreshes after a new category is created
- The newly created category becomes the selected one
- When the "done" button is pressed, the new selected category is saved (previously it was not).

* Pre-review cleanup

maxim-papier committed:
Fixed 2 lines long text label glitched in the `CategoryCellView`

maxim-papier committed:
Cleaned up an unused constant

Helvetii951 commented on TrackerApp/CoreData/CategoryStore.swift:
```diff
@@ -23,14 +21,15 @@ final class TrackerCategoryStore: NSObject {
-    /// Initialize and return fetchedResultsController
-    func fetchedResultsControllerForCategory() -> NSFetchedResultsController<CategoryData> {
+    // Initialize and return fetchedResultsController
+    func retrieveFetchedResultsController() -> NSFetchedResultsController<CategoryData> {
```
Лучше сделать `NSFetchedResultsController` [ленивым свойством](https://abhimuralidharan.medium.com/lazy-var-in-ios-swift-96c75cb8a13a). Так не нужно будет париться с `if-let` и `guard` внутри.

maxim-papier replied:
Спасибо! Сделано.

Helvetii951 commented on TrackerApp/Models/Category.swift:
```diff
@@ -0,0 +1,34 @@
+
+// MARK: - MOCKS
+
+extension Category {
+    static var mockCategory1: Self {
+        .init(name: "Спорт", trackers: [.mockCase1, .mockCase2,.mockCase3])
+    }
+    static var mockCategory2: Self {
+        .init(name: "Дом", trackers: [.mockCase4, .mockCase5])
+    }
+}
```
В этом спринте мы уже создаем "боевые" данные в слое БД. Моковые лучше убрать.

maxim-papier replied:
Да, действительно. Спасибо. Сделано.

Helvetii951 commented on TrackerApp/Scenes/Category (MVVM)/CategoryViewModel.swift:
```diff
@@ -0,0 +1,75 @@
+    
+    private var dependencies: DependencyContainer
+    
+    @Observable private(set) var categories: [CategoryData] = []
+    @Observable private(set) var buttonState: ButtonState = .add
+    @Observable private(set) var selectedCategory: CategoryData?
```
Биндинги хорошо реализованы👍

Рекомендую уже сейчас начинать знакомиться с реактивщиной. Концепция не очень простая, но супер сильная. Эти знания скоро будут необходимы в связке SwiftUI + Combine. Плюс такие знания станут преимуществом при трудоустройстве.

[Документация Apple по фреймворку Combine](https://developer.apple.com/documentation/combine)
[Неплохая вводная статья](https://habr.com/ru/companies/simbirsoft/articles/712332/)

maxim-papier replied:
Благодарю вас. Меня вообще эта тема очень будоражит. Хочу разбираться! Спасибо за статьи.

Helvetii951 commented on TrackerApp/Scenes/HomeVC.swift:
```diff
@@ -18,36 +22,21 @@ final class HomeVC: UITabBarController {
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+        
+        let shouldShowOnboarding = onboardingPersistenceService.hasCompletedOnboarding()
+        
+        if !shouldShowOnboarding {
+            let onboardingViewModel = OnboardingViewModel(onboardingPersistenceService: onboardingPersistenceService)
+            let onboardingVC = OnboardingView(viewModel: onboardingViewModel)
+            onboardingVC.modalPresentationStyle = .fullScreen
+            present(onboardingVC, animated: false, completion: nil)
+        }
```
Лучше такую логику делать в том месте, где принимается решение об установке корневого экрана. У нас это `SceneDelegate`.

maxim-papier replied:
Я пытался делать это в `SceneDelegate`, но так как я хочу онбординг показать поверх `HomeVC`, у меня выходит, что `HomeVC` ещё не создан, а `OnboardingView` уже показывается. Не нашёл красивого решения и поступил так.

maxim-papier commented on TrackerApp/Models/Category.swift:
```diff
@@ -0,0 +1,34 @@
+    }
+}
+
+extension Category: Sequence {
+    func makeIterator() -> IndexingIterator<[Tracker]> {
+        return trackers.makeIterator()
+    }
+}
```
Какой-то рудимент 😆. Спасибо. Убрал.

maxim-papier commented on TrackerApp/Scenes/Category (MVVM)/CategoryView.swift:
```diff
@@ -0,0 +1,158 @@
+
+final class CategoryView: UIViewController {
+    
+    private var dependencies: DependencyContainer
+    private var viewModel: CategoryViewModel
+    
+    weak var categorySelectionDelegate: CategorySelectionDelegate?
```
Большое спасибо. Статью изучу внимательно.

maxim-papier commented on TrackerApp/Scenes/Category (MVVM)/CategoryViewModel.swift:
```diff
@@ -0,0 +1,75 @@
+    case databaseError(String)
+}
+
+enum ButtonState {
+    case add
+    case ready
+}
```
Спасибо. Сделано.

maxim-papier committed:
Partially done PR #21 review feedback
- Made `fetchedResultsController` lazy
- Fixed dark theme when creating a new category
- Made `ButtonState` a nested type in `CategoryViewModel`
- Removed mock data from all models

Next step: I will implement the count of days when closing the tracker.

maxim-papier committed:
Tasks now close/reopen correctly and the number of times a task has been completed is accounted for. (#22)

maxim-papier commented:
Все исправления внёс. Спасибо за ревью!

maxim-papier committed:
Removed one force unwrap

maxim-papier committed:
Create TrackerApp.xcscheme

maxim-papier committed:
Merge branch 'main' into sprint_16
